cumulusci/tasks/bulkdata/tests/test_upsert.py

```diff
@@ -608,6 +608,166 @@ def test_simple_upsert__bad_key(
         assert "FirstName" in str(e.value)
         assert "update key" in str(e.value)
 
+    @responses.activate
+    def test_simple_upsert__smart(
+        self, create_task, cumulusci_test_repo_root, org_config
+    ):
+        domain = org_config.get_domain()
+        ver = CURRENT_SF_API_VERSION
+        responses.add(
+            method="PATCH",
+            url=f"https://{domain}/services/data/v{ver}/composite/sobjects/Contact/Email",
+            status=200,
+            json=[
+                {
+                    "id": "003P000001Y5exdIAB",
+                    "success": True,
+                    "errors": [],
+                    "created": True,
+                },
+                {
+                    "id": "003P000001Y5exXIAR",
+                    "success": True,
+                    "errors": [],
+                    "created": False,
+                },
+                {
+                    "id": "003P000001Y5exYIAR",
+                    "success": True,
+                    "errors": [],
+                    "created": False,
+                },
+            ],
+        )
+        responses.add(
+            method="POST",
+            url=f"https://{domain}/services/data/v{ver}/composite/sobjects",
+            status=200,
+            json=[{"id": "001J000002jcjvnIAA", "success": True, "errors": []}],
+        )
+        mock_describe_calls(domain=domain, version=CURRENT_SF_API_VERSION)
+        task = create_task(
+            LoadData,
+            {
+                "sql_path": cumulusci_test_repo_root
+                / "datasets/upsert/upsert_example_2.sql",
+                "mapping": cumulusci_test_repo_root
+                / "datasets/upsert/upsert_smart__native_field.yml",
+                "ignore_row_errors": True,
+                "set_recently_viewed": False,
+            },
+        )
+        task._update_credentials = mock.Mock()
+        task()
```

Comment:
Since we don't have any explicit asserts in these three tests, how about using the `responses` context-manager style so it will assert for us that all expected HTTP calls are made?

Reply:
I think I added the appropriate assertions in eb2f6b9. How does that look to you?

Reply:
Looks good to me.